Assignment3/Q1.py

```diff
@@ -0,0 +1,128 @@
+
+from Dictionary import Dictionary
+
+
+def create_corresponding_graph(rows, columns):
+    """Method takes two arguments and returns a adjacency list representation of the corresponding grid with given
+       number of rows and columns.
+
+        Args:
+            rows: integer, number of rows in the grid
+            columns: integer, number of columns in the grid
+
+        Returns:
+            a dictionary, adjacency list representation of the graph corresponding to the character grid given in
+            the problem. A vertex in a graph is an integer tuple which indicates the coordinates of an entry in
+            the grid.
+    """
+    # Using a dictionary data structure for the adjacency list representation of the graph.
+    graph = {}
+    for i in range(rows):
+        for j in range(columns):
+            # Add an entry (tuple of coordinates) for each cell in a grid.
+            graph[(i, j)] = []
+
+            # Create arrays of possible adjacent entries.
+            adjacent_horizontal = [i - 1, i, i + 1]
+            adjacent_vertical = [j - 1, j, j + 1]
+
+            # Populate each entry in a graph adjacency list representation.
+            for h in adjacent_horizontal:
+                for v in adjacent_vertical:
+                    # There are eight adjacent cells to each cell in a grid
+                    # Apart from edge cases, here we take care of edge cases.
+                    if 0 <= h < rows and 0 <= v < columns and not (h == i and v == j):
+                        graph[(i, j)] = graph[(i, j)] + [(h, v)]
+
+    # Return an adjacency list representation of a graph corresponding to a grid.
+    return graph
+
+
+def word_search(rows, columns, grid, dictionary):
+    """A method performs a word search on a character grid and returns a set of words (no duplicates) that can be formed
+       from grid's characters and are in a dictionary. Word search is case insensitive.
+
+        Args:
+            rows: an integer, number of rows in a grid
+            columns: an integer, number of columns in a grid
+            grid: a 2D array of characters, we are searching for words from dictionary in a grid
+            dictionary: a Dictionary object, contains the words to search for in a grid
+
+        Returns:
+            a set of words, complete list of words that can be formed from grid's characters and are in a dictionary,
+            there are no duplicated words in a set.
+    """
+    # Instantiate a list of words.
+    word_list = []
+
+    # Create a graph corresponding to the given grid.
+    graph = create_corresponding_graph(rows, columns)
+
+    # For each vertex in a graph perform a search from vertex to find all the words starting with corresponding
+    # character in a grid.
+    for key in graph:
+        words = word_search_recursive(dictionary, grid, graph, key)
+        word_list = word_list + words
+
+    # Remove duplicates from the word list.
+    word_list = set(word_list)
```

Comment:
As an optimization, you can work throughout with a `set` instead of a `list` so that, if there are many duplicates, you do not build a very long list, just to drop the duplicates later.